llvm/lib/Object/DXContainer.cpp

```diff
@@ -283,11 +287,35 @@ Error DirectX::RootSignature::parse(StringRef Data) {
       support::endian::read<uint32_t, llvm::endianness::little>(Current);
   Current += sizeof(uint32_t);
 
-  if (!dxbc::RootSignatureValidations::isValidRootFlag(FValue))
+  if (!dxbc::RootSignatureVerifier::verifyRootFlag(FValue))
     return validationFailed("unsupported root signature flag value read: " +
                             llvm::Twine(FValue));
   Flags = FValue;
 
+  assert(Current == Begin + RootParametersOffset);
+
+  ParametersHeaders.Data = Data.substr(
+      RootParametersOffset, NumParameters * sizeof(dxbc::RootParameterHeader));
+
+  ParameterSpaceOffset =
+      RootParametersOffset + NumParameters * sizeof(dxbc::RootParameterHeader);
+  size_t ParameterSpaceEnd =
+      Data.size() - ((NumStaticSamplers == 0) ? 0 : StaticSamplersOffset);
+
+  ParameterSpace = Data.substr(ParameterSpaceOffset,
+                               ParameterSpaceEnd - ParameterSpaceOffset);
+
+  for (auto P : param_header()) {
+
+    if (!dxbc::RootSignatureVerifier::verifyParameterType(P.ParameterType))
+      return validationFailed("unsupported parameter type value read: " +
```

Comment:
The container object file parser should not be doing semantic verification of the file. If the object file reader can structurally read the file, it should succeed even if the contents are semantically incorrect.

Reply:
I'm a little worried that the latest changes here may have thrown out too much validation.  For example, if we have garbage for a parameter type then we cannot structurally read the file since we need the parameter type to tell us how to interpret the rest of the file.  Without checking for that we run off the rails and crash, or we don't notice that we've ignored big chunks of the file and claim we successfully parsed it.